# Welcome to TapeAgents!


**TapeAgents** is a framework that leverages a structured, replayable log (**Tape**) of the agent session to facilitate all stages of the LLM Agent development lifecycle. In TapeAgents, the agent reasons by processing the tape and the LLM output to produce new thoughts, actions, control flow steps and append them to the tape. The environment then reacts to the agent’s actions by likewise appending observation steps to the tape.

In this tutorial, you will learn:
- how to create TapeAgents using the low-level API
- run and resume TapeAgents
- have one TapeAgent reuse another TapeAgent's tape as training data

In upcoming versions of this tutorial, you will also learn: 
- how to make a team TapeAgent with subagents
- how to build TapeAgents using available high-level APIs
- how to build a TapeAgent that streams partial steps

Other tutorials and examples will cover:
- code execution and browser use
- finetuning
- the TapeAgents apps (Studio and Browser)

# Setup
We're assuming that you already installed the project through the `make setup` or the jupyter notebook is running in the context of the project. If not, please refer to the [README](README.md) for more detailed instructions.

In [1]:
# Now set the OPENAI_API_KEY environment variable to your API key.

import os

if "OPENAI_API_KEY" not in os.environ:
    os.environ["OPENAI_API_KEY"] = "<your-api-key>"
    # os.environ["OPENAI_ORGANIZATION"] = "" # optional
today = "2024-09-17"  # fixed date for reproducible tests


# If you prefer to skip the OpenAI setup and not make any LLM calls, you can use ones from the cache.
# it will work instead of the real LLM fine as long as the prompts are not changed.
# Uncomment the following lines to use the cache:
#
# from tapeagents import llms
# import os
# llm_cache_path = "tests/res/intro_notebook/tapedata.sqlite"
# if not os.path.exists(llm_cache_path):
#     llm_cache_path = f"../{llm_cache_path}"
# assert os.path.exists(llm_cache_path)
# llms._REPLAY_SQLITE = llm_cache_path


# 1. Your first TapeAgent

In this section, we will build the simplest possible "hello world" agent. We will then go through all the new concepts that you need to know to understand the code. This section is quite long, but with the solid foundation you acquire here other TapeAgent tutorials will be easy to process.

Without further ado, here's the code!

In [2]:
from tapeagents.agent import Agent, Node
from tapeagents.core import Prompt, SetNextNode
from tapeagents.dialog_tape import AssistantStep, UserStep, DialogTape
from tapeagents.llms import LLMStream, LiteLLM
from tapeagents.prompting import tape_to_messages

llm = LiteLLM(model_name="gpt-4o-mini")


class MainNode(Node):
    name: str = "main"

    def make_prompt(self, agent: Agent, tape: DialogTape) -> Prompt:
        # Render the whole tape into the prompt, each step is converted to message
        return Prompt(messages=tape_to_messages(tape))

    def generate_steps(self, agent: Agent, tape: DialogTape, llm_stream: LLMStream):
        yield AssistantStep(content=llm_stream.get_text())  # Generate new step from the LLM output stream.
        yield SetNextNode(next_node="main")  # Which node to execute next, more on that later


agent = Agent[DialogTape].create(llm, nodes=[MainNode()])
start_tape = DialogTape(steps=[UserStep(content="Tell me about Vulcan in 3 sentences")])
final_tape = agent.run(start_tape).get_final_tape()  # agent will start executing the first node
print(f"Final tape: {final_tape.model_dump_json(indent=2)}")


Final tape: {
  "metadata": {
    "id": "ebb354e3-0aa5-43e1-bb15-ef38a9675595",
    "parent_id": "8bba3038-7999-4acd-bbb4-fde0599355f0",
    "author": "Agent",
    "author_tape_id": null,
    "n_added_steps": 2,
    "error": null,
    "result": {}
  },
  "context": null,
  "steps": [
    {
      "metadata": {
        "id": "a770fd96-47d8-4735-a47d-728805dd5396",
        "prompt_id": "",
        "node": "",
        "agent": "",
        "other": {}
      },
      "kind": "user",
      "content": "Tell me about Vulcan in 3 sentences"
    },
    {
      "metadata": {
        "id": "5cb3c0a4-02bd-4ac7-9c57-3fca095416ab",
        "prompt_id": "ffdd213b-a855-4601-8e7f-e233726bdc24",
        "node": "main",
        "agent": "Agent",
        "other": {}
      },
      "kind": "assistant",
      "content": "Vulcan is a fictional planet in the Star Trek universe, known as the homeworld of the Vulcan species, including the iconic character Spock. It is characterized by a rich history of logic, rea

Now let's learn about tapes, steps, prompts, llm streams, nodes and agents.

### Tape

The fundamental concept of the TapeAgents is the `Tape`, a comprehensive semantic level log of the agent's session. A `Tape` contains a context and a sequence of `Step` objects. As you can see, a TapeAgent runs by adding steps (such as `UserStep` or `AssistantStep`) to the _tape_. This example uses the `DialogTape` tape, which is a basic tape for user-assistant conversations. Let's see what are the possible steps in a `DialogTape`.

In [3]:
# We use Python generics to instantiate many different Tape types by
# specifying different Context and Step types. In the output of this cell,
# look at Union[UserStep, AssistantStep, ...]
# for the list of possible step types in the DialogTape.
DialogTape


tapeagents.core.Tape[Union[DialogContext, NoneType], Union[UserStep, ToolResult, SystemStep, AssistantThought, SetNextNode, Pass, Call, Respond, FinalStep, AssistantStep, ToolCalls]]

Some of these steps should be familiar to you. `UserStep`, `AssistantStep`, `SystemStep` and `ToolResult` correspond to `role=user`, `role=assistant`, `role=system` and `role=tool` LLM API messages respectively. `ToolCalls` and `AssistantThought` correspond to assistant messages where the LLM requests a tool call or produces an intermediate thought that is not meant to be shown to the user. `SetNextNode` and `Pass` are TapeAgent's internal step to control which node it should run at the next iteration (more on this below).

### Prompt format; LLMs

We use the industry-standard "chat.completions" prompt format in TapeAgents: a list of user/assistant/system/tool messages plus tool schemas.

In [4]:
# Almost all classes in TapeAgents are Pydantic base models.
# This allows easy validation, serialization and instrospection. For example,
# here we are able to list all the fields in the Prompt model.
Prompt.model_fields


{'id': FieldInfo(annotation=str, required=False, default_factory=<lambda>),
 'tools': FieldInfo(annotation=Union[list[dict], NoneType], required=False, default=None),
 'messages': FieldInfo(annotation=list[dict], required=False, default_factory=list)}

The LLMs in TapeAgent take `Prompt` and return an `LLMStream` object. The `LLMStream` object can be used both to fast-forward to the complete response text and to stream partial outputs step by step.

In [5]:
llm_stream = LiteLLM(model_name="gpt-4o-mini-2024-07-18", stream=True)

# Streaming
prompt = Prompt(messages=[{"role": "user", "content": "Write hello world in Java"}])
for event in llm_stream.generate(prompt):
    print(event.chunk, end="")

# No streaming
# (note: you can not use Prompt object for more than 1 LLM call in TapeAgents)
prompt = Prompt(messages=[{"role": "user", "content": "Write hello world in C"}])
print("\n" + "-" * 30)
print(llm_stream.generate(prompt).get_text())


Certainly! Here’s a simple Java program that prints "Hello, World!" to the console:

```java


public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!");
    }
}
```

To run this program, save it in a file named `HelloWorld.java`, then

 compile and run it using the following commands in a terminal:

1. Compile the program:
   ```
   javac HelloWorld.java
   ```

2.

 Run the compiled program:
   ```
   java HelloWorld
   ```

You should see the output:
```
Hello, World!
```None
------------------------------


Certainly! Here is the simplest "Hello, World!" program written in C:

```c
#include <stdio.h>

int main() {
    printf("Hello, World!\n");
    return 0;
}
```

To compile and run this program, you can use a C compiler like `gcc`. Here are the steps:

1. Save the code in a file named `hello.c`.
2. Open your terminal and navigate to the directory where the file is saved.
3. Compile the program with the command:
   ```
   gcc hello.c -o hello
   ```
4. Run the compiled program with the command:
   ```
   ./hello
   ```

You should see the output:
```
Hello, World!
```


In the example above we use the easiest way to create a prompt from the tapes: `tape_to_messages`. Under the hood, this method uses `step.llm_dict()` method of all non-control steps in the tape to create the prompt:

In [6]:
print((user := UserStep(content="hi AI!")).llm_dict())
print((assistant := AssistantStep(content="hello human")).llm_dict())
print(tape_to_messages(DialogTape(steps=[user, assistant])))


{'kind': 'user', 'content': 'hi AI!'}
{'kind': 'assistant', 'content': 'hello human'}
[{'role': 'user', 'content': 'hi AI!'}, {'role': 'assistant', 'content': 'hello human'}]


A key priority in TapeAgents is making use of the data that running the agent generates. To make this possible, some TapeAgent LLMs know how to make their finetuning data:

In [7]:
from tapeagents.core import LLMOutput
from tapeagents.llms import TrainableLLM

trainable_llm = TrainableLLM(
    base_url="",  # we only use the tokenizer from the model here, no need for a base_url for inference
    model_name="microsoft/Phi-3.5-MoE-instruct",
    tokenizer_name="microsoft/Phi-3.5-MoE-instruct",
)

simple_tape = DialogTape(
    steps=[
        UserStep(content="Say bla 3 times and foo 2 times"),
        AssistantStep(content="Sure! Let me say bla bla bla foo foo"),
    ]
)

prompt = Prompt(messages=tape_to_messages(simple_tape[:1]))  # type: ignore
output = agent.make_llm_output(simple_tape, index=1)
text = trainable_llm.make_training_text(prompt=prompt, output=output)
print("--- ALL TEXT ---")
print(text.text)
print("--- PREDICTED CHARACTERS ---")
print(text.output_text)


--- ALL TEXT ---
<|user|>
Say bla 3 times and foo 2 times<|end|>
<|assistant|>
Sure! Let me say bla bla bla foo foo<|end|>
<|endoftext|>
--- PREDICTED CHARACTERS ---
Sure! Let me say bla bla bla foo foo<|end|>
<|endoftext|>


### Node

A node represents an uninterruptible atom of TapeAgent's computation. When TapeAgents runs a node, it uses its two main functions: `make_prompt` to create an LLM Prompt from the tape and `generate_steps` to create new steps from the LLM output. To build a node, you can subclass `Node` and override these functions. Note that `generate_steps` must be a generator, a design choice we made to make TapeAgents a streaming-friendly framework. 

Let's see what the node from the above example can do.

In [8]:
from tapeagents.llms import LLMEvent


class MainNode(Node):
    name: str = "main"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        return Prompt(messages=tape_to_messages(tape))

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        yield AssistantStep(content=llm_stream.get_text())
        yield SetNextNode(next_node="main")  # Continue to the same node


node = MainNode()

# Let's run "make_prompt" in isolation.
prompt = node.make_prompt(agent=None, tape=DialogTape(steps=[UserStep(content="Hi, AI!")]))
print(f"Raw node prompt:\n{prompt}\n")


# Now, let's run "generate_steps" in isolation.
# We need to construct a fake LLMStream to do that.
def _generator():
    yield LLMEvent(output=LLMOutput(content="Hello, human!"))


stream = LLMStream(_generator(), Prompt())
step_stream = node.generate_steps(agent=None, tape=DialogTape(), llm_stream=stream)
print(f"Steps produced by node's generator:\n{list(step_stream)}\n")

# When the agent runs a node, it is equivalent to the following three steps:
# Step 1: make a prompt
start_tape = DialogTape(steps=[UserStep(content="Hi, AI!")])
prompt = node.make_prompt(agent, start_tape)
# Step 2: construct the LLMStream from the prompt (happens inside the agent)
stream = llm.generate(prompt)
# Step 3: generate steps that the agent will then add to the tape
print("Produced Steps:")
for step in node.generate_steps(agent, start_tape, stream):
    print(step.model_dump_json(indent=2))


Raw node prompt:
id='b074927f-c5f6-4881-809f-3aa7e0db8e89' tools=None messages=[{'role': 'user', 'content': 'Hi, AI!'}]

Steps produced by node's generator:
[AssistantStep(metadata=StepMetadata(id='b80b16f3-5ca8-4ac8-849b-649b07e91dce', prompt_id='', node='', agent='', other={}), kind='assistant', content='Hello, human!'), SetNextNode(metadata=StepMetadata(id='bb676628-2c9c-4242-b8d7-12f91fa30a26', prompt_id='', node='', agent='', other={}), kind='set_next_node', next_node='main')]

Produced Steps:


{
  "metadata": {
    "id": "37a1f660-13d6-48be-a4f5-f780bfb009af",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "assistant",
  "content": "Hello! How can I assist you today?"
}
{
  "metadata": {
    "id": "41ad4b7f-2259-4e40-a331-a5ec374e78d7",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "main"
}


### Agent and its nodes

The TapeAgent agent iteratively runs the nodes and appends the steps generated by each node to the tape. To select which next node to run, internally a TapeAgent computes the **tape view** object. The Tape remains the only **state** that the agent uses, the view only represents its content in a way that is convenient for the agent to use.


In [9]:
from tapeagents.view import TapeViewStack
from tapeagents.core import StepMetadata

# The "top" view in the tape view stack is the view of the current agent.
# Initially `top.last_node` is empty and the agent will run the first node from its list".
tape1 = DialogTape(steps=[UserStep(content="Hi, AI!")])
last_node = TapeViewStack.compute(tape1).top.last_node
print(f"1: {last_node}")
assert last_node == ""


# When the agent computes the view, it updates `top.last_node` with the node from the latest agent step
# The agent will search the next node after the last_node in its nodes list.
tape2 = DialogTape(
    steps=[
        UserStep(content="Hi, AI!"),
        AssistantStep(metadata=StepMetadata(prompt_id="123", node="main"), content="AI here, how I can help?"),
    ]
)
last_node = TapeViewStack.compute(tape2).top.last_node
print(f"2: {last_node}")
assert last_node == "main"

# The SetNextNode step on the tape changes `top.next_node` to the value of the `next_node` field in the SetNextNode step.
# The agent will use this value
tape3 = DialogTape(
    steps=[
        UserStep(content="Hi, AI!"),
        AssistantStep(metadata=StepMetadata(prompt_id="123"), content="AI here, how I can help?"),
        SetNextNode(next_node="act"),
    ]
)
next_node = TapeViewStack.compute(tape3).top.next_node
print(f"3: {next_node}")
assert next_node == "act"


1: 
2: main
3: act


By default the agent stops after the last node has produced an `Action` step. The action steps are the steps by which the agent requests information from the environment. For example, `AssistantStep` is an `Action` as it indicates the agent awaits the user response, `ToolCalls` is an action requesting tool call results. Let's look at all possible steps in `DialogTape` tape and see which of them are actions, observations and thoughts.

In [10]:
from tapeagents.core import Action, Pass, Thought, Observation
from tapeagents.dialog_tape import AssistantThought, ToolCalls, ToolResult

assert all([issubclass(step_class, Action) for step_class in [AssistantStep, ToolCalls]])
assert all([issubclass(step_class, Thought) for step_class in [AssistantThought, SetNextNode, Pass]])
assert all([issubclass(step_class, Observation) for step_class in [UserStep, ToolResult]])


Now we are ready to look at a simplified summary of the corner-stone `agent.run` algorithm.

1. Compute the new tape view
2. Choose the active agent (more on multi-agent TapeAgents later)
3. Choose the active node
4. Run the node and add steps on the tape
5. If the last node yielded an action, then stop, else repeat.

`agent.run` returns an `AgentStream` object which allows iterating through the agent's steps (or partial steps when streaming) and fast-forwardin to the complete new tape with `get_final_tape`.

#### Converse with a TapeAgent

Let's continue the conversation with the agent that previously responded to us with the Vulcan definition from the StarTrek.
Remember, the session is stored in the tape **final_tape**.

In [11]:
tape_to_continue = final_tape + [UserStep(content="No, I mean Vulcan the company")]
continued_tape = agent.run(tape_to_continue).get_final_tape()
print(continued_tape.model_dump_json(indent=2))


{
  "metadata": {
    "id": "760644a1-2ee5-4e72-b831-296706d838cb",
    "parent_id": "ed46a3d3-7738-401b-b7a9-716b751cc2d5",
    "author": "Agent",
    "author_tape_id": null,
    "n_added_steps": 2,
    "error": null,
    "result": {}
  },
  "context": null,
  "steps": [
    {
      "metadata": {
        "id": "a770fd96-47d8-4735-a47d-728805dd5396",
        "prompt_id": "",
        "node": "",
        "agent": "",
        "other": {}
      },
      "kind": "user",
      "content": "Tell me about Vulcan in 3 sentences"
    },
    {
      "metadata": {
        "id": "5cb3c0a4-02bd-4ac7-9c57-3fca095416ab",
        "prompt_id": "ffdd213b-a855-4601-8e7f-e233726bdc24",
        "node": "main",
        "agent": "Agent",
        "other": {}
      },
      "kind": "assistant",
      "content": "Vulcan is a fictional planet in the Star Trek universe, known as the homeworld of the Vulcan species, including the iconic character Spock. It is characterized by a rich history of logic, reason, and the

Note that the agent is able to continue talking to you thanks for `SetNextNode(next_node="main")` step that `generate_steps` produced. If you try to remove this step as an exercise, the agent will crash because there is only one node.

#### Tape rendering

LLM agents create a lot of data that can be overwhelming to process. In TapeAgents we render the tape with the associated prompts and outputs into a more readable HTML for you. To make this work, we store prompts and outputs in an SQLite database every time you call `agent.run()`.

Here's how to use tape rendering in the notebook:

In [12]:
from tapeagents.renderers import render_tape_with_prompts
from tapeagents.renderers.pretty import PrettyRenderer
from IPython.display import HTML

HTML(render_tape_with_prompts(continued_tape, PrettyRenderer()))


# 2. Your TapeAgent with planning and tools

Let's build a TapeAgent that plans and acts. We will be using OpenAI function calling capabilities in this example.

In [13]:
from tapeagents.core import SetNextNode
from tapeagents.dialog_tape import AssistantThought, ToolCalls
from tapeagents.environment import ToolEnvironment
from tapeagents.orchestrator import main_loop
from tapeagents.tools.stock import get_stock_data, get_stock_ticker

system_instruction = f"""
You will help the user to learn about financials of companies.
Use as many relevant tools as possible to include more details and facts in your responses.
Today is {today}.
"""
system_message = {"role": "system", "content": system_instruction}

env = ToolEnvironment([get_stock_ticker, get_stock_data])


class PlanNode(Node):
    name: str = "plan"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        guidance = "Write a natural language plan on how to use tools help the user. Output a list of numbered items, like 1., 2., 3., etc."
        guidance_message = {"role": "user", "content": guidance}
        return Prompt(
            messages=[system_message] + tape_to_messages(tape) + [guidance_message], tools=env.get_tool_schema_dicts()
        )

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        if content := llm_stream.get_output().content:
            yield AssistantThought(content=content)
        else:
            raise ValueError()


class ActNode(Node):
    name: str = "act"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        guidance = "Follow the plan you created to earlier. When you are done, respond to the user."
        guidance_message = {"role": "user", "content": guidance}
        return Prompt(
            messages=[system_message] + tape_to_messages(tape) + [guidance_message], tools=env.get_tool_schema_dicts()
        )

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        o = llm_stream.get_output()
        if o.content:
            yield AssistantStep(content=o.content)
            yield SetNextNode(next_node="plan")
        elif o.tool_calls:
            yield ToolCalls.from_llm_output(o)
            yield SetNextNode(next_node="act")
        else:
            raise ValueError()


agent1 = Agent.create(LiteLLM(model_name="gpt-4o", parameters={"temperature": 0.1}), nodes=[PlanNode(), ActNode()])

print("Run the agent!")
final_tape1 = None
for event in main_loop(agent1, DialogTape() + [UserStep(content="Tell me about Vulcan in 3 sentences")], env):
    if ae := event.agent_event:
        if ae.step:
            print(ae.step.model_dump_json(indent=2))
        if ae.final_tape:
            final_tape1 = ae.final_tape
    if event.observation:
        print(event.observation.model_dump_json(indent=2))
assert final_tape1
print("Final tape:")
HTML(render_tape_with_prompts(final_tape1, PrettyRenderer()))


Run the agent!


{
  "metadata": {
    "id": "a1e1eb84-2474-40b5-8112-bcf8ad483655",
    "prompt_id": "80efe780-0783-47ff-ad6e-f672072c8849",
    "node": "plan",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant_thought",
  "content": "1. Use the `functions.get_stock_ticker` tool to find the stock ticker symbol for Vulcan.\n2. Once the stock ticker is obtained, use the `functions.get_stock_data` tool to retrieve recent stock price data for Vulcan, specifying a suitable date range.\n3. Compile the information gathered from the stock ticker and stock data to provide a concise summary of Vulcan's recent financial performance."
}


{
  "metadata": {
    "id": "0d4e198e-5ff4-4051-a0b3-5b9925f49897",
    "prompt_id": "ec96de60-9c16-4688-b6f6-a39466cbdd3f",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "tool_calls": [
    {
      "function": {
        "name": "get_stock_ticker",
        "arguments": "{\"company_name\":\"Vulcan\"}"
      },
      "id": "call_RrUnjppmuNj1sBfumqHJL7VU",
      "type": "function"
    }
  ]
}
{
  "metadata": {
    "id": "b992cfe5-295d-4a9e-80b5-35040fe47b3e",
    "prompt_id": "ec96de60-9c16-4688-b6f6-a39466cbdd3f",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "act"
}


{
  "metadata": {
    "id": "f1a1e60f-b0ec-4d7e-bd8e-ed58d9af9035",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "tool",
  "content": "VMC",
  "tool_call_id": "call_RrUnjppmuNj1sBfumqHJL7VU"
}


{
  "metadata": {
    "id": "f617af01-17f5-424a-af83-ac36f9630b5f",
    "prompt_id": "b86ab091-aa34-42f0-9d86-206bbc1447f7",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "tool_calls": [
    {
      "function": {
        "name": "get_stock_data",
        "arguments": "{\"symbol\":\"VMC\",\"start_date\":\"2024-08-01\",\"end_date\":\"2024-09-16\"}"
      },
      "id": "call_fkUo1TYNjq6WVAiMs4AJ0bfR",
      "type": "function"
    }
  ]
}
{
  "metadata": {
    "id": "147193a5-2016-413c-bffc-658454db3221",
    "prompt_id": "b86ab091-aa34-42f0-9d86-206bbc1447f7",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "act"
}


{
  "metadata": {
    "id": "fe30b569-c25f-48f7-a8da-04889d78349d",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "tool",
  "content": [
    [
      "2024-07-31",
      274.510009765625
    ],
    [
      "2024-08-01",
      271.1300048828125
    ],
    [
      "2024-08-02",
      266.3399963378906
    ],
    [
      "2024-08-05",
      257.42999267578125
    ],
    [
      "2024-08-06",
      246.13999938964844
    ],
    [
      "2024-08-07",
      241.22000122070312
    ],
    [
      "2024-08-08",
      246.3300018310547
    ],
    [
      "2024-08-09",
      244.33999633789062
    ],
    [
      "2024-08-12",
      241.89999389648438
    ],
    [
      "2024-08-13",
      243.83999633789062
    ],
    [
      "2024-08-14",
      242.5
    ],
    [
      "2024-08-15",
      246.57000732421875
    ],
    [
      "2024-08-16",
      244.42999267578125
    ],
    [
      "2024-08-19",
      244.2899932861328
    ],
    [
      "2024-08-20",
     

{
  "metadata": {
    "id": "73ec7a51-6a16-4fc1-ada6-16e591f84bb0",
    "prompt_id": "1daa5386-8bc6-49c7-a3f9-0a656078b3f9",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "content": "Vulcan Materials Company, trading under the stock ticker symbol VMC, is a major producer of construction aggregates and materials. Over the past month and a half, Vulcan's stock price has experienced a decline, starting at approximately $274.51 on July 31, 2024, and decreasing to around $237.47 by September 13, 2024. This reflects a general downward trend in their stock value during this period, possibly influenced by broader market conditions or company-specific factors."
}
{
  "metadata": {
    "id": "ca38c607-904f-43e4-8e5c-edba786e1bf2",
    "prompt_id": "1daa5386-8bc6-49c7-a3f9-0a656078b3f9",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "plan"
}
Final tape:


The main new thing in this example is the environment. In TapeAgents framework the environment responds to the agent `Action` steps with `Observation` steps. We expect you to use the environment to encapsulate tool use, retrieval, code execution: everything that is non-deterministic, non-stationary, or computationally heavy. On the contrary, we encourage you to implements the agent's deterministic decision-making in `make_prompt` and `generate_steps` methods.

Here we use a pre-defined `main_loop` orchestrator to run the agent and the environment. `main_loop` is a generator of events that you can use as you wish. You are free to implement your own orchestration paradigm with a fine-grained control over what actions get to be executed.

# 3. Agent configuration, resumption

Let's try building a similar agent with an open-weights LLAMA3 70B models. Conveniently, [Together AI](together.ai) offers API endpoints. You can create an account and get API key with some free quota.

#### Access to Hugging Face Gated Models
Make sure you have access to each model we are going to use (read more [here](https://huggingface.co/docs/hub/en/models-gated#access-gated-models-as-a-user)):
- https://huggingface.co/meta-llama/Meta-Llama-3-70B-Instruct

##### Troubleshoot
- If you receive a 401 error, it means the authentication failed.
- If you receive a 403 error, it means you are authenticated but not authorized to access the specific model.

In [14]:
if "TAPEAGENTS_LLM_TOKEN" not in os.environ:
    os.environ["TAPEAGENTS_LLM_TOKEN"] = "<your-together-ai-api-key>"

if "HF_TOKEN" not in os.environ:
    # We need this to acces the model's tokenizer
    os.environ["HF_TOKEN"] = "<your-hugging-face-api-key>"


## 3.1. Setup your Agents

We've found that LLAMA3 function-calling is not yet battle-ready. We will use the structured output approach to make it call tools instead. We are also making this agent trainable by adding `make_llm_output` methods to each node. `Node.make_llm_output` defines how a node can reconstruct the LLM completion message that would be required to make the steps from the given tape at the given index. You can think of `Node.make_llm_output` as the inverse of `Node.generate_steps`.

When you run the code below, you might see a different behavior every time. Often the LLAMA-based agent gets stuck in a loop. We will look into how TapeAgents supports you in addressing this issue by
- tuning the prompt and resuming the agent exactly where it got stuck 
- producing training text from a different agent's tape 

In [15]:
import json
from tapeagents.dialog_tape import FunctionCall, ToolCall
from tapeagents.llms import TrainableLLM

from tapeagents.prompting import step_to_message

env = ToolEnvironment([get_stock_ticker, get_stock_data])

system_instruction = f"""
You will help the user to learn about financials of companies.
Use as many relevant tools as possible to include more details and facts in your responses.
Today is {today}.

You have access to the following tools: {env.get_tool_schema_dicts()}"""

planning_guidance = "Write a natural language plan on how to use tools help the user. Output a list of numbered items, like 1., 2., 3., etc."

call_or_respond_guidance = """
Follow the plan you created earlier. When you are done, respond to the user.
If you want to call a or several tools, output JSON like this
{"kind": "tool_call", "tool_name": "...", "parameters": "... unquoted parameters json ..."}
If you have called all the tools in the plan, respond to the user with the JSON of the form
{"kind": "response", "content": "... you response ... "}.
Output ONE JSON OBJECT ONLY PER LINE ONLY AND NOTHING ELSE.
"""


class PlanNode(Node):
    name: str = "plan"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        system_message = {"role": "system", "content": system_instruction}
        guidance_message = {"role": "user", "content": agent.templates["planning"]}
        return Prompt(messages=[system_message] + tape_to_messages(tape) + [guidance_message])

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        if content := getattr(llm_stream.get_output(), "content", None):
            yield AssistantThought(content=content)
        else:
            raise ValueError()

    def make_llm_output(self, agent, tape: DialogTape, index: int) -> LLMOutput:
        if not isinstance(current := tape[index], AssistantThought):
            raise ValueError()
        return LLMOutput(role="assistant", content=current.content)


class ActNode(Node):
    name: str = "act"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        system_message = {"role": "system", "content": system_instruction}
        guidance_message = {"role": "user", "content": agent.templates["call_or_respond"]}
        messages = [system_message]
        for step in tape:
            if isinstance(step, (ToolCalls)):
                messages.append({"role": "assistant", "content": _llm_message_content(step)})
            elif not isinstance(step, SetNextNode):
                messages.append(step_to_message(step))
        messages += [guidance_message]
        return Prompt(messages=messages)

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        o = llm_stream.get_output()
        try:
            assert o.content
            tool_calls = []
            response = None
            for line in o.content.split("\n"):
                data = json.loads(line)
                if data.get("kind") == "response":
                    response = data["content"]
                elif data.get("kind") == "tool_call":
                    tool_call = ToolCall(
                        function=FunctionCall(name=data["tool_name"], arguments=json.dumps(data["parameters"])),
                        # tool call must be a unique string, it helps to make it something deterministic
                        id=f"tool_call_{len(tool_calls)}_node_starts_at_{len(tape)}",
                    )
                    tool_calls.append(tool_call)
                else:
                    yield AssistantStep(content="Invalid LLM output: kind field must be 'response' or 'tool_call'")
            if response and tool_calls:
                yield AssistantStep(content="Invalid LLM output: response and tool_call cannot be in the same message")
            if response:
                yield AssistantStep(content=response)
                yield SetNextNode(next_node="plan")
            if tool_calls:
                yield ToolCalls(tool_calls=tool_calls)
                yield SetNextNode(next_node="act")
        except Exception as e:
            yield AssistantStep(content="Invalid JSON object: " + str(e))

    def make_llm_output(self, agent, tape: DialogTape, index: int) -> LLMOutput:
        if not isinstance(step := tape[index], AssistantStep | ToolCalls):
            raise ValueError()
        content = _llm_message_content(step)
        return LLMOutput(role="assistant", content=content)


def _llm_message_content(step: AssistantStep | ToolCalls):
    """Helper function to make both the prompt and the target completion"""
    match step:
        case AssistantStep():
            return json.dumps({"kind": "response", "content": step.content})
        case ToolCalls():
            content = ""
            for tool_call in step.tool_calls:
                if content:
                    content += "\n"
                content += json.dumps(
                    {
                        "kind": "tool_call",
                        "tool_name": tool_call.function.name,
                        "parameters": json.loads(tool_call.function.arguments),
                    }
                )
            return content
        case _:
            raise ValueError()


agent2 = Agent.create(
    TrainableLLM(
        base_url="https://api.together.xyz",
        model_name="meta-llama/Meta-Llama-3-70B-Instruct-Turbo",
        tokenizer_name="meta-llama/Meta-Llama-3-70B-Instruct",
        parameters=dict(temperature=0.01),
    ),
    templates={
        "system": system_instruction,
        "planning": planning_guidance,
        "call_or_respond": call_or_respond_guidance,
    },
    nodes=[PlanNode(), ActNode()],
)


## 3.2. Run your Agent

In [16]:
final_tape2 = None
print("Run LLAMA agent!")
for event in main_loop(
    agent2, DialogTape() + [UserStep(content="Tell me about Vulcan Materials in 3 sentences")], env, max_loops=3
):
    if ae := event.agent_event:
        if ae.step:
            print(ae.step.model_dump_json(indent=2))
        if ae.final_tape:
            final_tape2 = ae.final_tape
    if event.observation:
        print(event.observation.model_dump_json(indent=2))
assert final_tape2 is not None
print("Final tape:")
HTML(render_tape_with_prompts(final_tape2, PrettyRenderer()))


Run LLAMA agent!


{
  "metadata": {
    "id": "1e453f4f-d7d9-4b20-8516-5baaee5aefec",
    "prompt_id": "e9e46607-215c-4fc0-b298-0ed589d2b958",
    "node": "plan",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant_thought",
  "content": "Here's a plan to help the user learn about Vulcan Materials:\n\n1. Ask the user if they want to know the stock ticker of Vulcan Materials, and if so, use the `get_stock_ticker` function to retrieve the ticker symbol.\n\n2. Use the retrieved ticker symbol to get the stock price data of Vulcan Materials using the `get_stock_data` function, with a suitable date range (e.g., the last year or last quarter).\n\n3. Analyze the retrieved stock price data to provide the user with insights about Vulcan Materials' stock performance, such as the highest and lowest prices, average price, and any notable trends or fluctuations.\n\n4. Offer to provide additional information about Vulcan Materials, such as its industry, market capitalization, or recent news, and use extern

{
  "metadata": {
    "id": "56519040-bb6d-4cd8-b970-aacc7c5acdd9",
    "prompt_id": "9157181f-96fa-4a41-a44f-315fe01e5d79",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "tool_calls": [
    {
      "function": {
        "name": "get_stock_ticker",
        "arguments": "{\"company_name\": \"Vulcan Materials\"}"
      },
      "id": "tool_call_0_node_starts_at_2",
      "type": "function"
    }
  ]
}
{
  "metadata": {
    "id": "50d4cda4-bc10-49f5-bb52-7b1b1fa2d72a",
    "prompt_id": "9157181f-96fa-4a41-a44f-315fe01e5d79",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "act"
}
{
  "metadata": {
    "id": "965d5ed3-9fd2-4558-8b4f-c88beae23604",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "tool",
  "content": "VMC",
  "tool_call_id": "tool_call_0_node_starts_at_2"
}


{
  "metadata": {
    "id": "9614ba7d-079e-4680-ab58-9d6f67011661",
    "prompt_id": "d6ee8f7c-10dc-4700-81a0-7c8ac92b060e",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "tool_calls": [
    {
      "function": {
        "name": "get_stock_data",
        "arguments": "{\"symbol\": \"VMC\", \"start_date\": \"2023-09-17\", \"end_date\": \"2024-09-17\"}"
      },
      "id": "tool_call_0_node_starts_at_5",
      "type": "function"
    }
  ]
}
{
  "metadata": {
    "id": "79296d6f-cc9d-4a3d-b930-2bd45bdef45b",
    "prompt_id": "d6ee8f7c-10dc-4700-81a0-7c8ac92b060e",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "act"
}


{
  "metadata": {
    "id": "f7046db7-094e-4ac6-bd5a-4ac01e2cee9f",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "tool",
  "content": [
    [
      "2023-09-18",
      211.69000244140625
    ],
    [
      "2023-09-22",
      200.6199951171875
    ],
    [
      "2023-09-28",
      205.02999877929688
    ],
    [
      "2023-10-04",
      205.02999877929688
    ],
    [
      "2023-10-10",
      211.0
    ],
    [
      "2023-10-16",
      213.0399932861328
    ],
    [
      "2023-10-20",
      201.5500030517578
    ],
    [
      "2023-10-26",
      193.97000122070312
    ],
    [
      "2023-11-01",
      203.8000030517578
    ],
    [
      "2023-11-07",
      207.2100067138672
    ],
    [
      "2023-11-13",
      210.49000549316406
    ],
    [
      "2023-11-17",
      212.35000610351562
    ],
    [
      "2023-11-24",
      211.67999267578125
    ],
    [
      "2023-11-30",
      213.55999755859375
    ],
    [
      "2023-12-06",
    

{
  "metadata": {
    "id": "b0e36d98-1c2a-48f8-ac0f-a3c30859ea42",
    "prompt_id": "5a546adc-8a12-46d0-800e-65f84171a7f0",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "assistant",
  "tool_calls": [
    {
      "function": {
        "name": "get_stock_data",
        "arguments": "{\"symbol\": \"VMC\", \"start_date\": \"2023-09-17\", \"end_date\": \"2024-09-17\"}"
      },
      "id": "tool_call_0_node_starts_at_8",
      "type": "function"
    }
  ]
}
{
  "metadata": {
    "id": "182b788d-94d6-4d8d-903d-4edeb5f39a9b",
    "prompt_id": "5a546adc-8a12-46d0-800e-65f84171a7f0",
    "node": "act",
    "agent": "Agent",
    "other": {}
  },
  "kind": "set_next_node",
  "next_node": "act"
}
{
  "metadata": {
    "id": "30df7325-8dd4-4b75-93fe-19bdb10e1246",
    "prompt_id": "",
    "node": "",
    "agent": "",
    "other": {}
  },
  "kind": "tool",
  "content": [
    [
      "2023-09-18",
      211.69000244140625
    ],
    [
      "2023-09-22",
      200.61999511

Sometimes the above agent works well, but quite likely you are seeing that the LLAMA-based agent is having trouble. Let's try to help it. For reproducibility, we'll use a pre-recorded failed tape.

In [17]:
with open("assets/failed_tape.json") as src:
    failed_tape = DialogTape.model_validate(json.load(src))
agent2b = agent2.model_copy(deep=True)
agent2b.templates["call_or_respond"] += (
    "REMEMBER: check what tool calls you have already made. Do not do the same call again!"
)
resume_from_step8 = agent2b.run(failed_tape[:8]).get_final_tape()
HTML(render_tape_with_prompts(resume_from_step8, PrettyRenderer()))


We found that this helpful hint often gets LLAMA-based agent unstuck. Note how easy it was to test it thanks to the ability of the agent to resume from step 8!

# 4. Tape reuse and training data


Another way to help this agent (or one with an even smaller LLM) is to finetune the LLM. And the most important step towards finetuning is making the training data!

There are two ways to make training data in TapeAgents:
- the basic one: use the `LLMCall` structures that the agent created when it produced the tape. You can retrieve them from the SQLite storage and convert into training text.
- the much powerful one: call `agent.reuse` to reconstructed the prompts and outputs **and** to validate that with the reconstructed LLMCalls the agent would indeed create the given tape

The big advantage of the second approach is that it allows you to use the tape from a teacher agent (think slower and more expensive) to train a student agent (think faster and cheaper). Or to train an agent on its own revised tapes.

Of course, restrictions apply: the tape by agent A may not be reusable by agent B directly. You might have to add/remove some steps. But at least `agent.reuse` verifies if your tape modifications led to creation of a tape that the agent B can indeed produce.

#### Make training data from the past LLM calls

In [18]:
from tapeagents.observe import retrieve_tape_llm_calls

llm_calls = retrieve_tape_llm_calls(final_tape2)
print(f"Retrieved {len(llm_calls)} LLM calls from the tape.")
# under the hood agent2 will route this request to its llm
example_text = agent2.make_training_text(list(llm_calls.values())[0])
print("From the first retrieved LLM call, the LLM will be trained to predict this text:")
print("---")
print(example_text.output_text)


Retrieved 4 LLM calls from the tape.
From the first retrieved LLM call, the LLM will be trained to predict this text:
---
{"kind": "tool_call", "tool_name": "get_stock_data", "parameters": {"symbol": "VMC", "start_date": "2023-09-17", "end_date": "2024-09-17"}}<|eot_id|>


#### Make training data by reusing a tape

Note how `agent2` reuses the tape by `agent1`, even though they have very different prompt and output formats! 

You can inspect the reused tape below and see that the steps are the same as before, but the prompts and outputs are different.

In [19]:
reused_tape, _ = agent2.reuse(final_tape1)
HTML(render_tape_with_prompts(reused_tape, PrettyRenderer()))


We offer a quick way to harness the tape reuse to make training data. 

In [20]:
training_data = agent2.make_training_data(final_tape1)
print(training_data[0].output_text)


1. Use the `functions.get_stock_ticker` tool to find the stock ticker symbol for Vulcan.
2. Once the stock ticker is obtained, use the `functions.get_stock_data` tool to retrieve recent stock price data for Vulcan, specifying a suitable date range.
3. Compile the information gathered from the stock ticker and stock data to provide a concise summary of Vulcan's recent financial performance.<|eot_id|>


What could be simpler than that?!
We're only scratching the surface of what TapeAgents can do. We invite you to explore the other tutorials and examples to learn more about the TapeAgents framework.

# 5. Multi-agent teams, deeper dive into view stack

Let's add a colleague to our agent that will help it search the internet! 

First thing, we need to give this colleague some tools. In TapeAgents, the agents do not use the environment directly, their interaction with the environment is mediated by in orchestrator such as you application. But the agents should know what tools they can call. And if you are using `main_loop` as the orchestrator, it requires one environment that contains tools of all the agents. 

Let's go ahead and define environments:
- the environments of the root agent
- the one of its internet search specialist colleague
- the environment that contains all the tools.

Note that we won't use the first two environments to produce observations, we'll use them only to generate tool schemas for agent.

In [21]:
from tapeagents.tools.simple_browser import SimpleTextBrowser

browser = SimpleTextBrowser()
search_agent_env = ToolEnvironment([browser.get_search_results, browser.get_page, browser.get_next_page])


# We will use the tool choice mechanism to let the main agent call its search specialist agent.
# To this end, we create a mock tool that represents calling the search agent.
def call_search_agent(query: str):
    """Use this tool to ask a fellow AI agent to search for information on the web."""
    pass


main_agent_env = ToolEnvironment([get_stock_ticker, get_stock_data, call_search_agent])
whole_env = ToolEnvironment(
    [get_stock_ticker, get_stock_data, browser.get_search_results, browser.get_page, browser.get_next_page]
)


Before we implement the subagent, let's review the way we do multi-agent communication in TapeAgents:
- when the root agent wants to call its subagent "xyz", it puts `Call(agent_name="xyz")` step on the tape
- at the next iteration, the root agent will compute `TapeViewStack` and delegate to the currently active agent. Right after `Call(agent_name="B")` on top of the stack there will be a new view associated with Agent "xyz". The root agent will delegate to "xyz" to make the prompt and to generate the steps from "xyz"'s current node.
- when "xyz" is done it will put `Respond()` step on the tape. At the next iteration the view stack won't have "xyz"'s view on the top any more. 

A reader familiar with the concept of a call stack will find the `TapeViewStack` concept very similar...

Note that for the purpose of computing the view stack the root agent's name is not known. It's view will be signed as "root".

Let's explore the way `Call` and `Respond` works with a minimal example.


In [22]:
from tapeagents.core import Call, Respond

tape = DialogTape(
    steps=[
        UserStep(content="Compute 2 + 2"),
        Call(agent_name="xyz", content="what is 2 + 2, dear xyz"),
        AssistantThought(content="deep thinking by agent xyz"),
        Respond(content="I heard it is 4"),
    ]
)
# We will print a brief summary of view stack after each step
for i in range(0, len(tape)):
    print(f"View stack after step {i}")
    view_stack = TapeViewStack.compute(tape[: i + 1])
    for view in view_stack.stack:
        step_summary = ", ".join([step.__class__.__name__ for step in view.steps])
        print(f"-- {view.agent_full_name}: {step_summary}")
# Note how "root/xyz" view appears after step 1 and disappears after step 3.
# Also note how "root" does not see private thoughts of "xyz" (step 2), and "xyz"
# does not see the initial observation of root (step 0)


View stack after step 0
-- root: UserStep
View stack after step 1
-- root: UserStep, Call
-- root/xyz: Call
View stack after step 2
-- root: UserStep, Call
-- root/xyz: Call, AssistantThought
View stack after step 3
-- root: UserStep, Call, Respond


We are now ready to proceed to the implementation of the internet search agent!

In [23]:
from tapeagents.nodes import FixedStepsNode
from tapeagents.core import Respond
from tapeagents.prompting import view_to_messages

search_system_instruction = "Use at most 5 tool calls to search the request info on on the web."
search_system_message = {"role": "system", "content": search_system_instruction}


class SearchAgentMainNode(Node):
    name: str = "main"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        view = agent.compute_view(tape)
        return Prompt(messages=view_to_messages(view.top, agent), tools=search_agent_env.get_tool_schema_dicts())

    def generate_steps(self, agent, tape, llm_stream: LLMStream):
        o = llm_stream.get_output()
        if o.content:
            # if the LLM responds, yield Respond(..) as your last step
            yield Respond(content=o.content)
        elif o.tool_calls:
            # when the LLM suggests tool calls, yield them as action steps
            yield ToolCalls.from_llm_output(o)
            yield SetNextNode(next_node="main")
        else:
            raise ValueError()


search_agent = Agent.create(
    name="search_agent",
    llms=LiteLLM(model_name="gpt-4o", parameters={"temperature": 0.1}),
    nodes=[SearchAgentMainNode()],
)
# To test the subagent, we'll make a mock root agent that immediately calls "search_agent"
call_step = Call(agent_name="search_agent", content="What influenced Nvidia stock price in late 2022?")
test_root_agent = Agent.create(subagents=[search_agent], nodes=[FixedStepsNode(steps=[call_step])])
start_tape = DialogTape()
final_tape = None
for event in main_loop(test_root_agent, start_tape, search_agent_env):
    # We need to stop the loop when "search_agent" responds,
    # otherwise the code will crash because `test_root_agent` only has one node.
    if (ae := event.agent_event) and isinstance(ae.step, Respond):
        final_tape = ae.partial_tape
        break
assert final_tape
HTML(render_tape_with_prompts(final_tape, PrettyRenderer()))


Finally, let's add the search subagent to a financial analyst agent like the ones in earlier examples. We need to give the root agent a way to use the LLM to decide whether to call the search specialist, and if yes, what query to pass to it. We will abuse the tool calling mechanism for this purpose to make the example simpler.

In [24]:
from tapeagents.core import SetNextNode
from tapeagents.dialog_tape import AssistantThought, ToolCalls
from tapeagents.orchestrator import MainLoopStatus, main_loop
from tapeagents.view import Call
from IPython.display import clear_output

system_instruction = f"""
You will help the user to learn about financials of companies. 
For general user queries, include some info about stock price changes during the last year, as well as some general information on the company.
Today is {today}.
"""
system_message = {"role": "system", "content": system_instruction}


class PlanNode(Node):
    name: str = "plan"

    def make_prompt(self, agent, tape) -> Prompt:
        view = agent.compute_view(tape)
        guidance = "Write a natural language plan on how to use tools help the user. Output a list of numbered items, like 1., 2., 3., etc."
        guidance_message = {"role": "user", "content": guidance}
        return Prompt(
            messages=[system_message] + view_to_messages(view.top, agent) + [guidance_message],
            tools=main_agent_env.get_tool_schema_dicts(),
        )

    def generate_steps(self, agent, dialog, llm_stream: LLMStream):
        if content := llm_stream.get_output().content:
            yield AssistantThought(content=content)
        else:
            raise ValueError()


class ActNode(Node):
    name: str = "act"

    def make_prompt(self, agent, tape: DialogTape) -> Prompt:
        view = agent.compute_view(tape)
        guidance = "Follow the plan you created to earlier. When you are done, respond to the user."
        guidance_message = {"role": "user", "content": guidance}
        return Prompt(
            messages=[system_message] + view_to_messages(view.top, agent) + [guidance_message],
            tools=main_agent_env.get_tool_schema_dicts(),
        )

    def generate_steps(self, agent, dialog, llm_stream: LLMStream):
        o = llm_stream.get_output()
        if o.content:
            yield SetNextNode(next_node="plan")
            yield AssistantStep(content=o.content)
        elif o.tool_calls:
            yield SetNextNode(next_node="act")
            # only keep the tool calls before the call to another agent
            agent_call = None
            for i, tc in enumerate(o.tool_calls):
                if tc.function.name == "call_search_agent":
                    agent_call = tc
                    o.tool_calls = o.tool_calls[:i]
                    break
            # either produce the ToolCalls action OR call another agent
            if o.tool_calls:
                yield ToolCalls.from_llm_output(o)
            else:
                assert agent_call and agent_call.function.name == "call_search_agent"
                yield Call(agent_name="search_agent", content=json.loads(agent_call.function.arguments)["query"])

        else:
            raise ValueError()


multi_agent_analyst = Agent.create(
    name="analyst",
    subagents=[search_agent.clone()],
    llms=LiteLLM(model_name="gpt-4o", parameters={"temperature": 0.1}),
    nodes=[PlanNode(), ActNode()],
)

print("Run the agent!")
start_tape = DialogTape(steps=[UserStep(content="Tell me about Vulcan in 3 sentences")])
for event in main_loop(multi_agent_analyst, start_tape, whole_env):
    # This agent runs for a while, so we will show you a fresh render every time
    # when the environment finishes reacting with new actions
    if new_tape := event.agent_tape or event.env_tape:
        clear_output()
        display(HTML(render_tape_with_prompts(new_tape, PrettyRenderer())))
    # Uncomment this if you want to pause after every loop
    # if event.env_tape:
    #     input("Press Enter the run the next iteration of the main loop")
    if event.status == MainLoopStatus.EXTERNAL_INPUT_NEEDED:
        break


Look at this rather long tape and note how "analyst" calls "search_agent", and how the latter then responds to "analyst". Congratulations, you now know how to build a multi-agent TapeAgent!